app/controllers/technical_metadata_controller.rb

```diff
@@ -18,16 +19,67 @@ def show_by_druid
     head(:not_found) if @files.empty?
   end
 
+  # Given a druid (in the URL path) and a list of filename/md5 pairs (in the request body with filename URL encoded),
+  # return any differences between the expected (given) file signature list and the actual contents of technical
+  # metadata service's DB.
+  #
+  # @note This is just a read, and so should be idempotent, but we're using a POST instead of a GET, because
+  # the parameter list might exceed the traditional 2 kB limit on URL params for a GET request.
+  # @note The returned structure is 3 lists of hashes where each element has the same structure as the elements
+  # in the input list of hashes (i.e. 3 lists of filename/md5 pairs).  Matching with the input list is on
+  # filename. For the one field where the filename appears for the druid in both techMD service DB and the input list,
+  # mismatched_checksum_files, the md5 returned is what is recorded in techMD service's DB (as the caller already knows
+  # the checksum they provided).
+  #
+  # POST /v1/technical-metadata/audit/{druid}
+  def audit_by_druid # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+    files = DroFile.where(druid: params[:druid])
+    if files.empty?
+      head(:not_found) if files.empty?
+      return
+    end
+
+    recorded_files_for_comparison = files.map do |file|
+      { filename: file.filename, md5: file.md5 }
+    end
+
+    recorded_file_names = recorded_files_for_comparison.pluck(:filename)
+    missing_files = file_audit_infos.reject do |expected_file|
+      recorded_file_names.include?(expected_file[:filename])
+    end
+
+    expected_file_names = file_audit_infos.pluck(:filename)
+    unexpected_files = recorded_files_for_comparison.reject do |recorded_file|
+      expected_file_names.include?(recorded_file[:filename])
+    end
+
+    mismatched_checksum_files = file_audit_infos.filter_map do |expected_file|
+      recorded_files_for_comparison.select do |recorded_file|
+        recorded_file[:filename] == expected_file[:filename] && recorded_file[:md5] != expected_file[:md5]
+      end.presence # if no mismatches are found, []#presence will return nil, and filter_map will discard it
+    end.flatten
+
+    diff = { missing_files:, unexpected_files:, mismatched_checksum_files: }
+    render json: diff.to_json
```

Comment:
Some thought I had while reading this, which I will fully admit are somewhat rooted in MondayBrain... 🤔 :

* I wonder if some of these local var assignments could be private methods, given this is a hefty-ish controller action. But maybe that makes the action harder to read because these new concepts fly in without your eyes having seen them before?
* Would pulling this logic into a service class make sense, e.g., so it could be called in a rake task or in the console, or would that make it easier to test?
* I wonder if the two `Array#reject`-based calculations could be done using array difference to avoid iterations, *e.g.*: `missing_files = file_audit_infos.pluck(:filename) - recorded_files_for_comparison.pluck(:filename)`

But also, none of these are very clear or strong thoughts and I'd be happy to merge. 😄

Reply:
thanks for the suggestions!  all very good/reasonable.

> I wonder if the two `Array#reject`-based calculations could be done using array difference to avoid iterations, e.g.: `missing_files = file_audit_infos.pluck(:filename) - recorded_files_for_comparison.pluck(:filename)`

I couldn't figure out a way to do this that kept the overall changes simple, and didn't bloat the model class with something just for this PR, while also allowing me to return the filename/md5 pairs for the missing and unexpected files.  But then after some thinking, the md5 didn't seem particularly useful for those lists, so I ended up just returning lists of filenames for those scenarios.  That allowed me to use array difference as suggested and simplified things a good bit.

> I wonder if some of these local var assignments could be private methods, given this is a hefty-ish controller action. But maybe that makes the action harder to read because these new concepts fly in without your eyes having seen them before?

After that simplification, the mismatched files list was the only one with a knotty reject, so I ended up not doing further local var -> method extraction, since it felt less necessary, once the method was shorter.

> Would pulling this logic into a service class make sense, e.g., so it could be called in a rake task or in the console, or would that make it easier to test?

I ended up not going all the way on this, but I did extract the meat of the controller method logic into a private method that takes all it needs as params, so it'd be easy to further extract it into a service class.  I almost turned it to a class method, but that felt slightly less readable when I tried it.  I'll fully admit that this was partially post-AnchorWeek brain on my part, as I'd already had written the tests as request specs, and I didn't really want to rejigger those for a rake task that we don't yet have.  But happy to finish that extraction if you'd prefer.